sbi/inference/snpe/snpe_c.py

```diff
@@ -231,5 +370,354 @@ def _log_prob_proposal_posterior(
+        precisions_d: Tensor,
+    ):
+        """
+        Return the component weights (i.e. logits) of the proposal posterior.
+
+        Args:
+            means_pp: Means of the proposal posterior.
+            precisions_pp: Precision matrices of the proposal posterior.
+            covariances_pp: Covariance matrices of the proposal posterior.
+            logits_p: Component weights (i.e. logits) of the proposal distribution.
+            means_p: Means of the proposal distribution.
+            precisions_p: Precision matrices of the proposal distribution.
+            logits_d: Component weights (i.e. logits) of the density estimator.
+            means_d: Means of the density estimator.
+            precisions_d: Precision matrices of the density estimator.
+
+        Returns: Component weights of the proposal posterior. L*K terms.
+        """
+
+        num_comps_p = precisions_p.shape[1]
+        num_comps_d = precisions_d.shape[1]
+
+        # Compute alpha_i * beta_j
+        logits_p_rep = logits_p.repeat_interleave(num_comps_d, dim=1)
+        logits_d_rep = logits_d.repeat(1, num_comps_p)
+        logit_factors = logits_p_rep + logits_d_rep
+
+        # Compute sqrt(det()/(det()*det()))
+        covariances_p = torch.inverse(precisions_p)
+        covariances_d = torch.inverse(precisions_d)
+
+        logdet_covariances_pp = torch.logdet(covariances_pp)
+        logdet_covariances_p = torch.logdet(covariances_p)
+        logdet_covariances_d = torch.logdet(covariances_d)
+
+        # Repeat the proposal and density estimator terms such that there are LK terms.
+        # Same trick as has been used above.
+        logdet_covariances_p_rep = logdet_covariances_p.repeat_interleave(
+            num_comps_d, dim=1
+        )
+        logdet_covariances_d_rep = logdet_covariances_d.repeat(1, num_comps_p)
+
+        sqrt_det_ratio = 0.5 * (
+            logdet_covariances_pp
+            - (logdet_covariances_p_rep + logdet_covariances_d_rep)
+        )
+
+        # Compute for proposal, density estimator, and proposal posterior:
+        # mu_i.T * P_i * mu_i
+        exponent_p = _batched_vmv(precisions_p, means_p)
+        exponent_d = _batched_vmv(precisions_d, means_d)
+        exponent_pp = _batched_vmv(precisions_pp, means_pp)
+
+        # Extend proposal and density estimator exponents to get LK terms.
+        exponent_p_rep = exponent_p.repeat_interleave(num_comps_d, dim=1)
+        exponent_d_rep = exponent_d.repeat(1, num_comps_p)
+        exponent = -0.5 * (exponent_p_rep + exponent_d_rep - exponent_pp)
+
+        logits_pp = logit_factors + sqrt_det_ratio + exponent
+
+        return logits_pp
+
+    def _maybe_z_score_theta(self, theta: Tensor) -> Tensor:
+        """Return potentially standardized theta if z-scoring was requested."""
+
+        if self.z_score_theta:
+            theta, _ = self._posterior.net._transform(theta)
+
+        return theta
+
+
+def _batched_vmv(matrix: Tensor, vector: Tensor) -> Tensor:
```

Comment:
I think this could go into `sbiutils`

Reply:
I'd rather have it stay here. I can't image it being used for anything other than snpe-c (because it requires a dimension for `num_components`)

Do you have an application in mind?

Reply:
I'd rather keep it here. I can not think of an application outside of snpe-c (because it requires a dimension for num_components).

Or do you have sth in mind?

Reply:
The code itself does not depend on snpe c and is general, no? And it could be useful for SNPE B, for example.

Reply:
It is general, yes. But I do not directly see how it could be used for SNPE-B. Maybe for SNPE-A though.

Convinced :)

Reply:
I put it in `torchutils.py`, not in `sbiutils.py`. What do you think?

Reply:
Ah, reverting, putting it in `sbiutils.py` ;)